OIDC flow doc newbie improvements and big reorganization (#33794)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-code-flow-authentication-concept.adoc

= OpenID Connect authorization code flow mechanism for protecting web applications
include::_attributes.adoc[]
:categories: security,web
:toclevels: 4

> To protect your web applications, you can use the industry standard authorization code flow mechanism provided by the Quarkus OpenID Connect (OIDC) extension.

Reviewer: ```suggestion
To protect your web applications, you can use the industry-standard authorization code flow mechanism provided by the Quarkus OpenID Connect (OIDC) extension.
```

The following tokens are issued:

> * ID token: The Quarkus web-app uses the user information in the ID token to enable the authenticated user to log in securely and to provide role-based access to the web-app.

Reviewer 2: @michelle-purcell Lets keep it as `web-app` to avoid the confusion as it is an application type requiring a code flow authentication. `web app` is a general concept, we can have `web app` protected by the bearer token authentication or any other security mechanism

> * ID token: The Quarkus web-app uses the user information in the ID token to enable the authenticated user to log in securely and to provide role-based access to the web-app.

Reviewer 2: Also propose to have `application` instead of `component`

> * ID token: The Quarkus web-app uses the user information in the ID token to enable the authenticated user to log in securely and to provide role-based access to the web-app.

Reviewer: Thanks @ sberyozkin. In that case, how about ...? 

```suggestion
* ID token: The Quarkus `web-app` component uses the user information in the ID token to enable the authenticated user to log in securely and to provide role-based access to the web application.
```

== Using the authorization code flow mechanism

> === Configuring access to the OIDC Provider endpoint

Reviewer: Use sentence-case for titles, so only first word and proper nouns are capitalized.

```suggestion
=== Configuring access to the OIDC provider endpoint
```

> OIDC `web-app` application needs to know the endpoint URLs of OpenID Connect provider's authorization, token, `JsonWebKey` (JWK) set and possibly `UserInfo`, introspection and end session (RP-initiated logout).

Reviewer: Use abbreviation, to reduce wordiness unless it's the first time on the page or large section.

```suggestion
The OIDC `web-app` application requires the endpoint URLs of the OIDC provider's authorization, token, `JsonWebKey` (JWK) set, and possibly the `UserInfo`, introspection and end session (RP-initiated logout).
```

Reviewer 2: @michelle-purcell This sentence does not read right (your proposal is fine, but `endpoint` is in the wrong place), so I propose
```
The OIDC `web-app` application requires URLs of the OIDC provider's authorization, token, `JsonWebKey` (JWK) set, and possibly the `UserInfo`, introspection and end session (RP-initiated logout) endpoints.
```

Reviewer: Thanks for explaining this. I created a new suggestion with your rewrite.

By convention, they are discovered by adding a `/.well-known/openid-configuration` path to the configured `quarkus.oidc.auth-server-url`.

> Alternatively, if the discovery endpoint is not available, or if you would like to save on the discovery endpoint round-trip, you can disable the discovery and configure them with relative path values, for example:

Reviewer: ```suggestion
Alternatively, if the discovery endpoint is not available, or you prefer to reduce the discovery endpoint round-trip, you can disable endpoint discovery and configure relative path values, for example:
```

[source, properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
quarkus.oidc.discovery-enabled=false
# Authorization endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/auth
quarkus.oidc.authorization-path=/protocol/openid-connect/auth
# Token endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/token
quarkus.oidc.token-path=/protocol/openid-connect/token
# JWK set endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/certs
quarkus.oidc.jwks-path=/protocol/openid-connect/certs
# UserInfo endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/userinfo
quarkus.oidc.user-info-path=/protocol/openid-connect/userinfo
# Token Introspection endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/token/introspect
quarkus.oidc.introspection-path=/protocol/openid-connect/token/introspect
# End session endpoint: http://localhost:8180/realms/quarkus/protocol/openid-connect/logout
quarkus.oidc.end-session-path=/protocol/openid-connect/logout
----

> Sometimes, your OpenId Connect provider supports a metadata discovery but does not return all the endpoint URLs required for the authorization code flow to complete or for the application to support the additional functions such as a user logout. In such cases, you can simply configure missing endpoints URL locally:

Reviewer: ```suggestion
Some OIDC providers support metadata discovery but do not return all the endpoint URLS required for the authorization code flow to complete or to support application functions, for example, user logout. 
To work around this limitation, you can configure the missing endpoint URLs locally, as outlined in the following example:
```

Reviewer 2: @michelle-purcell `URLS` does not read right, it reads like an entirely new acronym, may be `endpoint URL values` ?

Reviewer: @sberyozkin - Great catch :+1:  I edited my comment suggestion to fix.

[source, properties]
----
# Metadata is auto-discovered but it does not return an end-session endpoint URL

quarkus.oidc.auth-server-url=http://localhost:8180/oidcprovider/account

# Configure the end-session URL locally, it can be an absolute or relative (to 'quarkus.oidc.auth-server-url') address
quarkus.oidc.end-session-path=logout
----

> Exactly the same configuration can be used to override a discovered endpoint URL if that URL does not work for the local Quarkus endpoint and a more specific value is required. For example, one can imagine that in the above example, a provider which supports both global and application specific end-session endpoints returns a global end-session URL such as `http://localhost:8180/oidcprovider/account/global-logout` which will logout the user from all the applications this user is currently logged in, while the current application only wants to get this user logged out from this application, therefore, `quarkus.oidc.end-session-path=logout` is used to override the global end-session URL.

Reviewer: ```suggestion
You can use this same configuration to override a discovered endpoint URL if that URL does not work for the local Quarkus endpoint and a more specific value is required. 
For example, a provider that supports both global and application-specific end-session endpoints returns a global end-session URL such as `http://localhost:8180/oidcprovider/account/global-logout`. 
This URL will log the user out of all of the applications that the user is currently logged into.
However, if the requirement is for the current application to log the user out of only the specific application, you can override the global end-session URL, by setting the `quarkus.oidc.end-session-path=logout` parameter.
```

[[oidc-provider-client-authentication]]
==== OIDC provider client authentication

> OIDC providers typically require using applications to be identified and authenticated when they interact with the OIDC endpoints.

Reviewer 2: @michelle-purcell Good catch, propose to drop `using`

> Quarkus OIDC (specifically `quarkus.oidc.runtime.OidcProviderClient`) authenticates to the OpenID Connect Provider when the authorization code has to be exchanged for the ID, access and refresh tokens, when the ID and access tokens have to be refreshed or introspected.

Reviewer 2: @michelle-purcell please replace `component` with `class`

> Quarkus OIDC (specifically `quarkus.oidc.runtime.OidcProviderClient`) authenticates to the OpenID Connect Provider when the authorization code has to be exchanged for the ID, access and refresh tokens, when the ID and access tokens have to be refreshed or introspected.

Reviewer: @sberyozkin - Done ^^.

> Typically, client id and client secrets are defined for a given application when it enlist to the OIDC provider.

Reviewer 2: ```suggestion
Typically, client id and client secrets are defined for a given application when it enlists to the OIDC provider.
```

Author: Applied

> Typically, client id and client secrets are defined for a given application when it enlists to the OIDC provider.

Reviewer: ```suggestion
Typically, client ID and client secrets are defined for a given application when it enlists to the OIDC provider.
```

Reviewer 2: @michelle-purcell `client_id` is a known OIDC property, `id` is usually used in the lower case, so if you don't mind lets keep it as is

Reviewer 2: That said, it reads OK, so let me commit it

> All the https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication[OIDC Client Authentication] options are supported, for example:

Reviewer: ```suggestion
All the link:https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication[OIDC Client] authentication options are supported, for example:
```

> `client_secret_basic`:

Reviewer: .Example of `client_secret_basic`

Reviewer 2: @michelle-purcell Please created a suggested change that I can commit

Reviewer: Done!

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.secret=mysecret
----

or

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.client-secret.value=mysecret
----

> or with the secret retrieved from a xref:credentials-provider.adoc[CredentialsProvider]:

Reviewer: ```suggestion
The following example shows the secret retrieved from a xref:credentials-provider.adoc[credentials provider]:
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app

# This is a key which will be used to retrieve a secret from the map of credentials returned from CredentialsProvider
quarkus.oidc.credentials.client-secret.provider.key=mysecret-key
# Set it only if more than one CredentialsProvider can be registered
quarkus.oidc.credentials.client-secret.provider.name=oidc-credentials-provider
----

> `client_secret_post`:

Reviewer: ```suggestion
.Example of `client_secret_post`
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.client-secret.value=mysecret
quarkus.oidc.credentials.client-secret.method=post
----

> `client_secret_jwt`, signature algorithm is HS256:

Reviewer: ```suggestion
.Example of `client_secret_jwt`, where the signature algorithm is HS256:
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.jwt.secret=AyM1SysPpbyDfgZld3umj1qzKObwVMkoqQ-EstJQLr_T-1qS0gZH75aKtMN3Yj0iPS4hcgUuTwjAzZr1Z9CAow
----

> or with the secret retrieved from a xref:credentials-provider.adoc[CredentialsProvider]:

Reviewer: ```suggestion
.Example of `client_secret_jwt`, where the secret is retrieved from a xref:credentials-provider.adoc[credentials provider]:
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app

# This is a key which will be used to retrieve a secret from the map of credentials returned from CredentialsProvider
quarkus.oidc.credentials.jwt.secret-provider.key=mysecret-key
# Set it only if more than one CredentialsProvider can be registered
quarkus.oidc.credentials.jwt.secret-provider.name=oidc-credentials-provider
----

> `private_key_jwt` with the PEM key file, signature algorithm is RS256:

Reviewer: ```suggestion
Example of `private_key_jwt` with the PEM key file, and where the signature algorithm is RS256:
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.jwt.key-file=privateKey.pem
----

> `private_key_jwt` with the keystore file, signature algorithm is RS256:

Reviewer: ```suggestion
.Example of `private_key_jwt` with the keystore file, where the signature algorithm is RS256:
```

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.jwt.key-store-file=keystore.jks
quarkus.oidc.credentials.jwt.key-store-password=mypassword
quarkus.oidc.credentials.jwt.key-password=mykeypassword

# Private key alias inside the keystore
quarkus.oidc.credentials.jwt.key-id=mykeyAlias
----

> Using `client_secret_jwt` or `private_key_jwt` authentication methods ensures that no client secret goes over the wire.

Reviewer: Is 'over the wire' a globally recognized term in security? Will it translate easily for all languages? If not, rephrase.

```suggestion
Using the `client_secret_jwt` or `private_key_jwt` authentication methods ensures that no client secret is leaked to an unauthorized entity.
```
or some similar term that's more technically correct.

Reviewer 2: @michelle-purcell I propose instead of `goes over the wire`, to have `is not sent to the OIDC provider, thus avoiding a risk of this secret being intercepted using a 'man-in-the-middle' attack`, that would be more technically correct.

Reviewer: @sberyozkin - I rewrote the sentence in the suggestion above. WDYT?

===== Additional JWT authentication options

> If `client_secret_jwt`, `private_key_jwt` authentication methods are used or an Apple `post_jwt` method is used, then the JWT signature algorithm, key identifier, audience, subject and issuer can be customized, for example:

Reviewer: Spell out JWT on first use if the audience needs it.

Reviewer 2: Hi @michelle-purcell I think we can assume that by this point the users are already aware of what `JWT` is, it is a very well known acronym, but indeed, we can spell it out in the doc introductions if it is not already done, this section is fairly low level/technical

[source,properties]
----
# private_key_jwt client authentication

quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.jwt.key-file=privateKey.pem

# This is a token key identifier 'kid' header - set it if your OpenID Connect provider requires it.
# Note if the key is represented in a JSON Web Key (JWK) format with a `kid` property then
# using 'quarkus.oidc.credentials.jwt.token-key-id' is not necessary.
quarkus.oidc.credentials.jwt.token-key-id=mykey

# Use RS512 signature algorithm instead of the default RS256
quarkus.oidc.credentials.jwt.signature-algorithm=RS512

# The token endpoint URL is the default audience value, use the base address URL instead:
quarkus.oidc.credentials.jwt.audience=${quarkus.oidc-client.auth-server-url}

# custom subject instead of the client id :
quarkus.oidc.credentials.jwt.subject=custom-subject

# custom issuer instead of the client id :
quarkus.oidc.credentials.jwt.issuer=custom-issuer
----

===== Apple POST JWT

> Apple OpenID Connect Provider uses a `client_secret_post` method where a secret is a JWT produced with a `private_key_jwt` authentication method but with Apple account specific issuer and subject claims.

Reviewer: ```suggestion
The Apple OIDC provider uses a `client_secret_post` method whereby a secret is a JWT produced with a `private_key_jwt` authentication method but with the Apple account-specific issuer and subject claims.
```

> `quarkus-oidc` supports a non-standard `client_secret_post_jwt` authentication method that can be configured as follows:

Reviewer: ```suggestion
In Quarkus Security, `quarkus-oidc` supports a non-standard `client_secret_post_jwt` authentication method, which you can configure as follows:
```

[source,properties]
----
# Apple provider configuration sets a 'client_secret_post_jwt' authentication method
quarkus.oidc.provider=apple

quarkus.oidc.client-id=${apple.client-id}
quarkus.oidc.credentials.jwt.key-file=ecPrivateKey.pem
quarkus.oidc.credentials.jwt.token-key-id=${apple.key-id}
# Apple provider configuration sets ES256 signature algorithm

quarkus.oidc.credentials.jwt.subject=${apple.subject}
quarkus.oidc.credentials.jwt.issuer=${apple.issuer}
----

> ===== Mutual TLS (mTLS)

Reviewer: ```suggestion
===== mutual TLS (mTLS)
```

> Some OpenID Connect providers might require that a client is authenticated as part of the `Mutual TLS` authentication process.

Reviewer: ```suggestion
Some OpenID Connect providers might require that a client is authenticated as part of the mutual TLS  authentication process.
```

`quarkus-oidc` can be configured as follows to support `mTLS`:

[source,properties]
----
quarkus.oidc.tls.verification=certificate-validation

# Keystore configuration
quarkus.oidc.tls.key-store-file=client-keystore.jks
quarkus.oidc.tls.key-store-password=${key-store-password}

# Add more keystore properties if needed:
#quarkus.oidc.tls.key-store-alias=keyAlias
#quarkus.oidc.tls.key-store-alias-password=keyAliasPassword

# Truststore configuration
quarkus.oidc.tls.trust-store-file=client-truststore.jks
quarkus.oidc.tls.trust-store-password=${trust-store-password}
# Add more truststore properties if needed:
#quarkus.oidc.tls.trust-store-alias=certAlias
----

==== Introspection endpoint authentication

Some OpenID Connect Providers may require authenticating to its introspection endpoint using Basic authentication with the credentials different to `client_id` and `client_secret` which may have already been configured to support `client_secret_basic` or `client_secret_post` client authentication methods described in the <<oidc-provider-client-authentication, OIDC provider client authentication>> section.

> If the tokens have to be introspected and the introspection endpoint specific authentication mechanism is required, then you can configure `quarkus-oidc` like this:

Reviewer: ```suggestion
If the tokens have to be introspected and the introspection endpoint-specific authentication mechanism is required, then you can configure `quarkus-oidc`, as follows:
```

[source,properties]
----
quarkus.oidc.introspection-credentials.name=introspection-user-name
quarkus.oidc.introspection-credentials.secret=introspection-user-secret
----

==== Redirecting to and from the OIDC provider

> When the user is redirected to the OpenID Connect Provider to authenticate, the redirect URL includes a `redirect_uri` query parameter which indicates to the provider where the user has to be redirected to once the authentication has been completed.

Reviewer: ```suggestion
When a user is redirected to the OpenID Connect provider to authenticate, the redirect URL includes a `redirect_uri` query parameter, which indicates to the provider where the user has to be redirected to when the authentication is complete.
```

In our case, this is the Quarkus application.

> Quarkus will set this parameter to the current application request URL by default. For example, if the user is trying to access a Quarkus service endpoint at `http://localhost:8080/service/1` then the `redirect_uri` parameter will be set to `http://localhost:8080/service/1`. Similarly, if the request URL is `http://localhost:8080/service/2` then the `redirect_uri` parameter will be set to `http://localhost:8080/service/2`, etc.

Reviewer: ```suggestion
Quarkus sets this parameter to the current application request URL by default. 
For example, if a user is trying to access a Quarkus service endpoint at `http://localhost:8080/service/1` then the `redirect_uri` parameter is set to `http://localhost:8080/service/1`. 
Similarly, if the request URL is `http://localhost:8080/service/2` then the `redirect_uri` parameter is set to `http://localhost:8080/service/2`.
```

> Some OpenID Connect Providers require the `redirect_uri` to have the same value for a given application (e.g. `http://localhost:8080/service/callback`) for all the redirect URLs.

Reviewer: ```suggestion
Some OIDC providers require the `redirect_uri` to have the same value for a given application, for example, `http://localhost:8080/service/callback`, for all the redirect URLs.
```

> In such cases, a `quarkus.oidc.authentication.redirect-path` property has to be set, for example, `quarkus.oidc.authentication.redirect-path=/service/callback`, and Quarkus will set the `redirect_uri` parameter to an absolute URL such as `http://localhost:8080/service/callback` which will be the same regardless of the current request URL.

Reviewer: ```suggestion
In such cases, a `quarkus.oidc.authentication.redirect-path` property has to be set, for example, `quarkus.oidc.authentication.redirect-path=/service/callback`, and Quarkus will set the `redirect_uri` parameter to an absolute URL such as `http://localhost:8080/service/callback`, which will be the same regardless of the current request URL.
```

> If `quarkus.oidc.authentication.redirect-path` is set but

Reviewer: Should this be joined with the next sentence? 

```suggestion
If `quarkus.oidc.authentication.redirect-path` is set but you need the original request URL to be restored after the user has been redirected back to a unique callback URL, for example,  `http://localhost:8080/service/callback`, set `quarkus.oidc.authentication.restore-path-after-redirect` property to `true`.
```

> If you need the original request URL to be restored after the user has been redirected back to a unique callback URL such as `http://localhost:8080/service/callback`, set `quarkus.oidc.authentication.restore-path-after-redirect` property to `true`.

Reviewer: ```suggestion
```

This will restore the request URL such as `http://localhost:8080/service/1`.

==== Customizing authentication requests

By default, only the `response_type` (set to `code`), `scope` (set to 'openid'), `client_id`, `redirect_uri` and `state` properties are passed as HTTP query parameters to the OpenID Connect provider's authorization endpoint when the user is redirected to it to authenticate.

> You can add more properties to it with `quarkus.oidc.authentication.extra-params`. For example, some OpenID Connect providers may choose to return the authorization code as part of the redirect URI's fragment which would break the authentication process - it can be fixed as follows:

Reviewer: ```suggestion
You can add more properties to it with `quarkus.oidc.authentication.extra-params`. 
For example, some OIDC providers might choose to return the authorization code as part of the redirect URI's fragment, which would break the authentication process.
The following example shows how you can work around this issue:
```

[source,properties]
----
quarkus.oidc.authentication.extra-params.response_mode=query
----

==== Customizing the authentication error response

> If the user authentication has failed at the OpenID Connect Authorization endpoint, for example, due to an invalid scope or other invalid parameters included in the redirect to the provider, then the provider will redirect the user back to Quarkus not with the `code` but `error` and `error_description` parameters.

Reviewer: ```suggestion
If the user authentication fails at the OIDC authorization endpoint, then the provider will redirect the user back to Quarkus with `error` and `error_description` parameters instead of `code`.
For example, this can happen when an invalid scope or other invalid parameters are included in the redirect to the provider.
```

In such cases HTTP `401` will be returned by default. However, you can instead request that a custom public error endpoint is called in order to return a user-friendly HTML error page. Use `quarkus.oidc.authentication.error-path`, for example:

[source,properties]
----
quarkus.oidc.authentication.error-path=/error
----

> It has to start from a forward slash and be relative to the base URI of the current endpoint.

Reviewer: ```suggestion
Ensure that the property starts with a forward slash (/) character and the path is relative to the base URI of the current endpoint.
```

For example, if it is set as '/error' and the current request URI is `https://localhost:8080/callback?error=invalid_scope` then a final redirect will be made to `https://localhost:8080/error?error=invalid_scope`.

> It is important that this error endpoint is a public resource to avoid the user redirected to this page be authenticated again.

Reviewer 2: @michelle-purcell The message here is that if this error page is not a public resource then Quarkus will need to initiate the authentication again when the user lands on the page after being redirected to it, so I think the original statement is correct

> It is important that this error endpoint is a public resource to avoid the user redirected to this page be authenticated again.

Reviewer: @sberyozkin - :+1:  I restored the original with just a teeny-tiny grammar update.

> It is important that this error endpoint is a public resource to avoid the user redirected to this page be authenticated again.

Reviewer 2: Cool

=== Accessing authorization data

Let's first discuss how to access information around authorization.

[[access_id_and_access_tokens]]
==== Accessing ID and access tokens
//SJ: new concept topic to describe the different token types and usage proposed in next iteration
The OIDC code authentication mechanism acquires three tokens during the authorization code flow: https://openid.net/specs/openid-connect-core-1_0.html#IDToken[IDToken], Access Token, and Refresh Token.

ID Token is always a JWT token and is used to represent a user authentication with the JWT claims.

> You can for example get which OIDC enpoint issued it, the user name and other information called _claims_.

Reviewer: ```suggestion
You can use this to get the issuing OIDC endpoint, the username, and other information called _claims_.
```

== Using the authorization code flow mechanism

=== Configuring access to the OIDC provider endpoint

> OIDC `web-app` application needs to know the endpoint URLs of OpenID Connect provider's authorization, token, `JsonWebKey` (JWK) set and possibly `UserInfo`, introspection and end session (RP-initiated logout).

Reviewer: @sberyozkin :+
```suggestion
The OIDC `web-app` application requires URLs of the OIDC provider's authorization, token, `JsonWebKey` (JWK) set, and possibly the `UserInfo`, introspection, and end session (RP-initiated logout) endpoints.
```

[source, properties]
----
# Metadata is auto-discovered but it does not return an end-session endpoint URL

quarkus.oidc.auth-server-url=http://localhost:8180/oidcprovider/account

# Configure the end-session URL locally, it can be an absolute or relative (to 'quarkus.oidc.auth-server-url') address
quarkus.oidc.end-session-path=logout
----

You can use this same configuration to override a discovered endpoint URL if that URL does not work for the local Quarkus endpoint and a more specific value is required. 
For example, a provider that supports both global and application-specific end-session endpoints returns a global end-session URL such as `http://localhost:8180/oidcprovider/account/global-logout`. 
This URL will log the user out of all of the applications that the user is currently logged into.
However, if the requirement is for the current application to log the user out of only the specific application, you can override the global end-session URL, by setting the `quarkus.oidc.end-session-path=logout` parameter.

[[oidc-provider-client-authentication]]
==== OIDC provider client authentication

> OIDC providers typically require using applications to be identified and authenticated when they interact with the OIDC endpoints.

Reviewer: ```suggestion
OIDC providers typically require applications to be identified and authenticated when they interact with the OIDC endpoints.
```

}
----

Note that `AccessTokenCredential` is used if the Access Token issued to the Quarkus `web-app` application is opaque (binary) and can not be parsed to a `JsonWebToken` or if the inner content is necessary for the application.

Injection of the `JsonWebToken` and `AccessTokenCredential` is supported in both `@RequestScoped` and `@ApplicationScoped` contexts.

> RefreshToken is used (by Quarkusn OIDC) to refresh the current ID and access tokens as part of its <<session-management,session management>> process.

Reviewer: ```suggestion
RefreshToken is used by Quarkusn OIDC to refresh the current ID and access tokens as part of its <<session-management,session management>> process.
```

Set the `quarkus.oidc.authentication.user-info-required=true` property for a https://openid.net/specs/openid-connect-core-1_0.html#UserInfo[UserInfo] JSON object from the OIDC userinfo endpoint to be requested.

A request will be sent to the OpenID Provider UserInfo endpoint using the access token returned with the authorization code grant response and an `io.quarkus.oidc.UserInfo` (a simple `jakarta.json.JsonObject` wrapper) object will be created. `io.quarkus.oidc.UserInfo` can be either injected or accessed as a SecurityIdentity `userinfo` attribute.

[[config-metadata]]
==== Accessing the OIDC configuration information

The current tenant's discovered link:https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata[OpenID Connect Configuration Metadata] is represented by `io.quarkus.oidc.OidcConfigurationMetadata` and can be either injected or accessed as a `SecurityIdentity` `configuration-metadata` attribute.

The default tenant's `OidcConfigurationMetadata` is injected if the endpoint is public.

[[token-claims-roles]]

> ==== Mapping Token claims and SecurityIdentity roles

Reviewer: ```suggestion
==== Mapping token claims and `SecurityIdentity` roles
```

The way the roles are mapped to the SecurityIdentity roles from the verified tokens is identical to how it is done for the xref:security-oidc-bearer-token-authentication-concept.adoc[Bearer tokens] with the only difference being that https://openid.net/specs/openid-connect-core-1_0.html#IDToken[ID Token] is used as a source of the roles by default.

[NOTE]

> --

Reviewer: (FYI... other note markup isn't supported downstream, hence the change.)
```suggestion
====
```

> If you use Keycloak then you should set a `microprofile_jwt` client scope for ID token to contain a `groups` claim, please see the https://www.keycloak.org/docs/latest/server_admin/#protocol[Keycloak Server Administration Guide] for more information.

Reviewer: ```suggestion
If you use Keycloak, set a `microprofile_jwt` client scope for ID token to contain a `groups` claim.
For more information, see the link:https://www.keycloak.org/docs/latest/server_admin/#protocol[Keycloak Server Administration Guide].
```

> --

Reviewer: ```suggestion
====
```

But depending on your OIDC provider, roles might be stored in the access token or the user info.

If the access token contains the roles and this access token is not meant to be propagated to the downstream endpoints then set `quarkus.oidc.roles.source=accesstoken`.

If UserInfo is the source of the roles then set `quarkus.oidc.roles.source=userinfo`, and if needed, `quarkus.oidc.roles.role-claim-path`.

Additionally, a custom `SecurityIdentityAugmentor` can also be used to add the roles. For more information, see xref:security-customization.adoc#security-identity-customization[SecurityIdentity customization].

=== Ensuring validity of tokens and authentication data

> A core part of the authentication process is ensuring the chain of trust and validity of the information.

Reviewer 2: Nice, useful to have an explanation like that one before going straight to the code/configuration :+1:

This is done by ensuring tokens are trustable.

[[token-verification-introspection]]

> ==== Token verification and introspection

Reviewer 2: Sure, `====` was nearly an only option when everything, the quickstart demo description and all of this content squashed in one doc, but now making these sections more visible would help

> OIDC inherits the bearer authentication token verification and introspection logic described in detail at xref:security-oidc-bearer-token-authentication-concept.adoc#token-verification-introspection[Token Verification And Introspection].

Reviewer 2: ```suggestion
Verification process of OIDC authorization code flow tokens follows the bearer authentication token verification and introspection logic described in detail at xref:security-oidc-bearer-token-authentication-concept.adoc#token-verification-introspection[Token Verification And Introspection].
```

Reviewer 2: The bearer token verification is very often done to verify the access tokens posted by SPAs which implement OIDC authorization code flow :-). DevServices for Keycloak UI is a good example, it is an SPA which does the code flow, posts the tokens as bearer tokens to Quarkus.
It would be useful indeed to descrive the relationship between OIDC and OAuth2 though, but in a dedicated section, here I feel `inherits` might be a bit confusing. I can look at adding a dedicated section

Author: OK. I'd advise aginst going into details of OIDC vs OAuth2, it's not the purpose of that doc.

Author: Applied

> Verification process of OIDC authorization code flow tokens follows the bearer authentication token verification and introspection logic described in detail at xref:security-oidc-bearer-token-authentication-concept.adoc#token-verification-introspection[Token Verification And Introspection].

Reviewer 2: @michelle-purcell My understanding is that we usually don't follow up the link name with the name of the guide.
Can we simply have the link name renamed from `Token Verification And Introspection` to `Bearer authentication token verification and introspection` ?

> Verification process of OIDC authorization code flow tokens follows the bearer authentication token verification and introspection logic described in detail at xref:security-oidc-bearer-token-authentication-concept.adoc#token-verification-introspection[Token Verification And Introspection].

Reviewer: @sberyozkin - We agreed during the Quarkus style discussions/definition phase that where possible we will try to use the above approach for external references. I'll check/update the contributor guide if needed.

- Specify the full title of the guide and use sentence case.
- Add the definite article “the” and “Quarkus” before the title. 
- Exception:  Omit the “Quarkus” prefix if “Quarkus is already in the title. For example, “see the Quarkus “Quarkus Application Data” guide.”
- Specify  “guide” directly after the title. Do not include “guide” as part of the hyperlink itself unless it is explicitly part of the title.

The benefits of this approach include clarity and context, especially if the deep link becomes available in the future, the reader can use the context to find the guide as needed. 
Another benefit is that this is also the approach used downstream, so fewer edits / diverging downstream.

Can we keep this?

[NOTE]
--

> In case of `web-app` applications, only the `IdToken` is verified by default since the access token is not used to access the current Quarkus `web-app` endpoint and rather meant to be propagated to the services expecting this access token.

Reviewer: ```suggestion
In case of `web-app` applications, only the `IdToken` is verified by default because the access token is not used to access the current Quarkus `web-app` endpoint and is intended to be propagated to the services expecting this access token.
```

If you expect the access token to contain the roles required to access the current Quarkus endpoint (`quarkus.oidc.roles.source=accesstoken`) then it will also be verified.
--

[[token-introspection-userinfo-cache]]
==== Token introspection and UserInfo cache

Code flow access tokens are not introspected unless they are expected to be the source of roles.

> It will however be used to get `UserInfo`.

Reviewer: ```suggestion
They will however be used to get `UserInfo`.
```

There will be one or two remote calls with the code flow access token, if the token introspection and/or `UserInfo` are required.

Please see xref:security-oidc-bearer-token-authentication-concept.adoc#token-introspection-userinfo-cache[Token Introspection and UserInfo cache] for more information about using a default token cache or registering a custom cache implementation.

[[jwt-claim-verification]]
==== JSON web token claim verification

Please see xref:security-oidc-bearer-token-authentication-concept.adoc#jwt-claim-verification[JSON Web Token Claim verification] section about the claim verification, including the `iss` (issuer) claim.
It applies to ID tokens but also to access tokens in a JWT format if the `web-app` application has requested the access token verification.

> ==== Further security with Proof key for code exchange (PKCE)

Reviewer: ```suggestion
==== Further security with Proof Key for Code Exchange (PKCE)
```

> link:https://datatracker.ietf.org/doc/html/rfc7636[Proof Key for Code Exchange] (PKCE) minimizes the risk of the authorization code interception.

Reviewer: ```suggestion
link:https://datatracker.ietf.org/doc/html/rfc7636[Proof Key for Code Exchange] (PKCE) minimizes the risk of authorization code interception.
```

> While `PKCE` is of primary importance to the public OpenID Connect clients (such as the SPA scripts running in a browser), it can also provide an extra level of protection to Quarkus OIDC `web-app` applications which are confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.

Reviewer: ```suggestion
While `PKCE` is of primary importance to public OpenID Connect clients, such as SPA scripts running in a browser, it can also provide an extra level of protection to Quarkus OIDC `web-app` applications. 
Quarkus OIDC `web-app` applications are confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.
```

Reviewer 2: @michelle-purcell It feels like that describing that `web-app` applications are confidential clients in a separate sentence kind of loses the link with the `PKCE` in this context. Looking at again, I wonder if we can just replace `which` with `even though` to make it clearer that PKCE is not strictly essential for this type of the application:
```
While `PKCE` is of primary importance to public OpenID Connect clients, such as SPA scripts running in a browser, it can also provide an extra level of protection to Quarkus OIDC `web-app` applications even though they act as confidential OpenID Connect clients capable of securely storing the client secret and using it to exchange the code for the tokens.
```

What do you think ?

Reviewer: I updated my suggestion above. WDYT?

> You can enable `PKCE` for your OIDC `web-app` endpoint with a `quarkus.oidc.authentication.pkce-required` property and a 32 characters long secret, for example:

Reviewer: ```suggestion
You can enable `PKCE` for your OIDC `web-app` endpoint with a `quarkus.oidc.authentication.pkce-required` property and a 32-character secret, as shown in the following example:
```

[source, properties]
----
quarkus.oidc.authentication.pkce-required=true
quarkus.oidc.authentication.pkce-secret=eUk1p7UB3nFiXZGUXi0uph1Y9p34YhBU
----

> If you already have a 32 characters long client secret then `quarkus.oidc.authentication.pkce-secret` does not have to be set unless you prefer to use a different secret key.

Reviewer: ```suggestion
If you already have a 32-characters client secret then you do not need to set the  `quarkus.oidc.authentication.pkce-secret` property unless you prefer to use a different secret key.
```

> The secret key is required for encrypting a randomly generated `PKCE` `code_verifier` while the user is being redirected with the `code_challenge` query parameter to OpenID Connect Provider to authenticate. The `code_verifier` will be decrypted when the user is redirected back to Quarkus and sent to the token endpoint alongside the `code`, client secret and other parameters to complete the code exchange. The provider will fail the code exchange if a `SHA256` digest of the `code_verifier` does not match the `code_challenge` provided during the authentication request.

Reviewer: ```suggestion
The secret key is required for encrypting a randomly generated `PKCE` `code_verifier` while the user is being redirected with the `code_challenge` query parameter to an OIDC provider to authenticate.
The `code_verifier` is decrypted when the user is redirected back to Quarkus and sent to the token endpoint alongside the `code`, client secret, and other parameters to complete the code exchange. 
The provider will fail the code exchange if a `SHA256` digest of the `code_verifier` does not match the `code_challenge` that was provided during the authentication request.
```

> === Handling and controlling the life time of authentication

Reviewer: ```suggestion
=== Handling and controlling the lifetime of authentication
```
or 

```suggestion
=== Handling and controlling the session lifetime of authentication
```

> Another important aspect of authentication is to be sure that data it is based on is fresh while not requiring the user to authenticate at every single request.

Reviewer: ```suggestion
Another important requirement for authentication is to ensure that the data the session is based on is up-to-date without requiring the user to authenticate for every single request.
```

> There are also situation when a logout is explicitly requested.

Reviewer: ```suggestion
There are also situations when a logout event is explicitly requested.
```

> This section discusses the various key notions necessarity to find the right balance for your Quarkus application.

Reviewer: Would active voice help here?

```suggestion
Consider the following security lifespan factors when securing your Quarkus application:
```

Reviewer 2: Hi @michelle-purcell I think this change changes a bit the meaning of what Emmanuel is saying, `lifespan` is only one of the technical properties of the token. 
```
This section discusses the various key notions necessary to find the right balance for securing your Quarkus application.
```
should work better ?

Reviewer: Where possible, we tried to avoid self-referential statements, especially when action is required.

How about ...? 
```suggestion
Use the following key points to find the right balance for securing your Quarkus applications:
```

Reviewer 2: @michelle-purcell This is perfect, thanks

[[oidc-cookies]]
==== Cookies

The OIDC adapter uses cookies to keep the session, code flow, and post-logout state.

> This state is a key element controlling the life time of authentication data.

Reviewer: ```suggestion
This state is a key element controlling the lifetime of authentication data.
```

`quarkus.oidc.authentication.cookie-path` property is used to ensure the same cookie is visible when you access the protected resources with overlapping or different roots, for example:

* `/index.html` and `/web-app/service`
* `/web-app/service1` and `/web-app/service2`
* `/web-app1/service` and `/web-app2/service`

> `quarkus.oidc.authentication.cookie-path` is set to `/` by default but can be narrowed to a more specific root path such as `/web-app`.

Reviewer: ```suggestion
By default, `quarkus.oidc.authentication.cookie-path` is set to `/` but you can change this to a more specific path if required, for example, `/web-app`.
```

Sometimes the cookie path needs to be set dynamically.
Set the `quarkus.oidc.authentication.cookie-path-header` property.

> For example, setting `quarkus.oidc.authentication.cookie-path-header=X-Forwarded-Prefix` means that the value of HTTP `X-Forwarded-Prefix` header will be used to set a cookie path.

Reviewer: ```suggestion
For example, to generate the cookie path dynamically by using the value of the`X-Forwarded-Prefix` HTTP header, set the property to `quarkus.oidc.authentication.cookie-path-header=X-Forwarded-Prefix.
```

Reviewer 2: @michelle-purcell Propose to replace `generate` with `set` as `generate` implies some randomness

Reviewer: Updated the suggestion. :+1:

If `quarkus.oidc.authentication.cookie-path-header` is set but no configured HTTP header is available in the current request then the `quarkus.oidc.authentication.cookie-path` will be checked.

> If your application is deployed across multiple domains, set a `quarkus.oidc.authentication.cookie-domain` property for the session cookie be visible to all protected Quarkus services, for example, if you have 2 services deployed at:

Reviewer: ```suggestion
If your application is deployed across multiple domains, set the `quarkus.oidc.authentication.cookie-domain` property so that the session cookie is visible to all protected Quarkus services.
For example, if you have two Quarkus services deployed at the following domains:
```

* https://whatever.wherever.company.net/
* https://another.address.company.net/

> then the `quarkus.oidc.authentication.cookie-domain` property must be set to `company.net`.

Reviewer: Make active:
```suggestion
Set the `quarkus.oidc.authentication.cookie-domain` property to `company.net`.
```

Reviewer 2: @michelle-purcell Can we have `then set` instead of `Set` ? Otherwise it is a new sentence which kind of loses the connection to the example

Reviewer: @sberyozkin - I see what you mean. I updated my comment with an alternative way that includes "..then set..." in the sentence.

I also added the `\` to the domain examples to escape them, otherwise, they get flagged as broken links on the Red Hat portal.

[[token-state-manager]]
===== Customizing the cookie with TokenStateManager

> OIDC `CodeAuthenticationMechanism` is using the default `io.quarkus.oidc.TokenStateManager` interface implementation to keep the ID, access and refresh tokens returned in the authorization code or refresh grant responses in a session cookie. It makes Quarkus OIDC endpoints completely stateless.

Reviewer: ```suggestion
OIDC `CodeAuthenticationMechanism` uses the default `io.quarkus.oidc.TokenStateManager` interface implementation to keep the ID, access, and refresh tokens returned in the authorization code or to refresh grant responses in a session cookie.
This makes Quarkus OIDC endpoints completely stateless.
```

> Note that some endpoints do not require the access token. An access token is only required if the endpoint needs to retrieve `UserInfo` or access the downstream service with this access token or use the roles associated with the access token (the roles in the ID token are checked by default). In such cases you can set either `quarkus.oidc.token-state-manager.strategy=id-refresh-token` (keep ID and refresh tokens only) or `quarkus.oidc.token-state-manager.strategy=id-token` (keep ID token only).

Reviewer: One sentence per line. 
Also consider using a list to reduce wordiness:

```suggestion
An access token is only required if the endpoint needs to:

* Retrieve `UserInfo`
* Access the downstream service with this access token
* Use the roles associated with the access token, which are checked by default

In such cases, use the `quarkus.oidc.token-state-manager.strategy` property to configure the token state strategy as follows:
 
[Attributes]
|===
|To... |Set the property to ... 
  
|Keep the ID and refresh tokens only
|`quarkus.oidc.token-state-manager.strategy=id-refresh-token` 

|Keep the ID token only 
|`quarkus.oidc.token-state-manager.strategy=id-token`

|===

```

> If the ID, access and refresh tokens are JWT tokens then combining all of them (if the strategy is the default `keep-all-tokens`) or only ID and refresh tokens (if the strategy is `id-refresh-token`) may produce a session cookie value larger than 4KB and the browsers may not be able to keep this cookie.

Reviewer: ```suggestion
If your chosen cookie strategy combines tokens and generates a large session cookie value that is greater than 4KB, some browsers might not be able to handle such cookie sizes.
This can occur when the ID, access, and refresh tokens are JWT tokens and the selected strategy is `keep-all-tokens` or with ID and refresh tokens when the strategy is `id-refresh-token`.
```

> In such cases, you can use `quarkus.oidc.token-state-manager.split-tokens=true` to have a unique session token per each of these tokens.

Reviewer: ```suggestion
To workaround this issue,  you can set `quarkus.oidc.token-state-manager.split-tokens=true` to generate a unique session token for each token.
```

Reviewer 2: @michelle-purcell `generate` implies some randomness which does not present in this context, so may be `create` ? Usually the cookies are `created` so it should be fine

Reviewer: Fixed :+1:

> Note that `TokenStateManager` will encrypt the tokens before storing them in the session cookie.

Reviewer: ```suggestion
`TokenStateManager` encrypts the tokens before storing them in the session cookie.
```

> For example, here is how you configure it to split the tokens and encrypt them:

Reviewer: ```suggestion
The following example shows how you configure `TokenStateManager` to split the tokens and encrypt them:
```

[source, properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
quarkus.oidc.client-id=quarkus-app
quarkus.oidc.credentials.secret=secret
quarkus.oidc.application-type=web-app
quarkus.oidc.token-state-manager.split-tokens=true
quarkus.oidc.token-state-manager.encryption-secret=eUk1p7UB3nFiXZGUXi0uph1Y9p34YhBU
----

> The token encryption secret must be at least 32 characters long. If this key is not configured then either `quarkus.oidc.credentials.secret` or `quarkus.oidc.credentials.jwt.secret` will be hashed to create an encryption key.

Reviewer: ```suggestion
The token encryption secret must be at least 32 characters long. 
If this key is not configured then either `quarkus.oidc.credentials.secret` or `quarkus.oidc.credentials.jwt.secret` will be hashed to create an encryption key.
```

> `quarkus.oidc.token-state-manager.encryption-secret` should be configured if Quarkus authenticates to OpenId Connect Provider using either mTLS or `private_key_jwt` (where a private RSA or EC key is used to sign a JWT token) authentication methods, otherwise a random key will be generated which will be problematic if the Quarkus application is running in the cloud with multiple pods managing the requests.

Reviewer: Reduce wordiness:

```suggestion

Configure the `quarkus.oidc.token-state-manager.encryption-secret` property if Quarkus authenticates to the OpenId Connect Provider by using one of the following authentication methods:

* mTLS
* `private_key_jwt`, where a private RSA or EC key is used to sign a JWT token

Otherwise, a random key is generated, which can be problematic if the Quarkus application is running in the cloud with multiple pods managing the requests.
```

If you need you can disable encrypting the tokens in the session cookie with `quarkus.oidc.token-state-manager.encryption-required=false`.

> Register your own `io.quarkus.oidc.TokenStateManager' implementation as an `@ApplicationScoped` CDI bean if you need to customize the way the tokens are associated with the session cookie. For example, you may want to keep the tokens in a database and have only a database pointer stored in a session cookie. Note though that it may present some challenges in making the tokens available across multiple microservices nodes.

Reviewer: One sentence per line:

```suggestion
Register your own `io.quarkus.oidc.TokenStateManager' implementation as an `@ApplicationScoped` CDI bean if you need to customize the way the tokens are associated with the session cookie. 
For example, you may want to keep the tokens in a database and have only a database pointer stored in a session cookie. 
Note that this approach might introduce some challenges if you need to make the tokens available across multiple microservices nodes.
```

* `/index.html` and `/web-app/service`
* `/web-app/service1` and `/web-app/service2`
* `/web-app1/service` and `/web-app2/service`

By default, `quarkus.oidc.authentication.cookie-path` is set to `/` but you can change this to a more specific path if required, for example, `/web-app`.

To set the cookie path dynamically, configure the `quarkus.oidc.authentication.cookie-path-header` property.
Set the `quarkus.oidc.authentication.cookie-path-header` property.
For example, setting `quarkus.oidc.authentication.cookie-path-header=X-Forwarded-Prefix` means that the value of HTTP `X-Forwarded-Prefix` header will be used to set a cookie path.

If `quarkus.oidc.authentication.cookie-path-header` is set but no configured HTTP header is available in the current request then the `quarkus.oidc.authentication.cookie-path` will be checked.

If your application is deployed across multiple domains, set the `quarkus.oidc.authentication.cookie-domain` property so that the session cookie is visible to all protected Quarkus services.

> For example, if you have two Quarkus services deployed at the following domains:

Reviewer: ```suggestion
For example, if you have Quarkus services deployed on the following two domains, then you must set the `quarkus.oidc.authentication.cookie-domain` property to `company.net`.

* \https://whatever.wherever.company.net/
* \https://another.address.company.net/
```

* https://whatever.wherever.company.net/
* https://another.address.company.net/

> then the `quarkus.oidc.authentication.cookie-domain` property must be set to `company.net`.

Reviewer: ```suggestion
```

> Note that a cookie name will vary when using xref:security-openid-connect-multitenancy.adoc[OpenID Connect Multi-Tenancy]. For example, it will be named `q_post_logout_tenant_1` for a tenant with a `tenant_1` ID, and so on.

Reviewer: ```suggestion
Note that a cookie name varies when using xref:security-openid-connect-multitenancy.adoc[OpenID Connect Multi-Tenancy]. For example, it will be named `q_post_logout_tenant_1` for a tenant with a `tenant_1` ID, and so on.
```

You may also need to set `quarkus.oidc.authentication.cookie-path` to a path value common to all the application resources which is `/` in this example.
For more information, see the <<oidc-cookies, Cookies>> section.

[NOTE]

> --

Reviewer: ```suggestion
====
```

> Some OpenID Connect providers do not support https://openid.net/specs/openid-connect-session-1_0.html#RPLogout[RP-Initiated Logout] specification (possibly because it is still technically a draft) and do not return an OpenID Connect well-known `end_session_endpoint` metadata property. However, it is not be a problem since these providers' specific logout mechanisms only differ in how the logout URL query parameters are named.

Reviewer: Remove assumptions (probably...) that might not be true in the near future.

```suggestion
Some OIDC providers do not support link:https://openid.net/specs/openid-connect-session-1_0.html#RPLogout[RP-initiated logout] specification and do not return an OpenID Connect well-known `end_session_endpoint` metadata property. 
However, this is not a problem for Quarkus because the specific logout mechanisms of such OIDC providers only differ in how the logout URL query parameters are named.
```

> According to the https://openid.net/specs/openid-connect-session-1_0.html#RPLogout[RP-Initiated Logout] specification, the `quarkus.oidc.logout.post-logout-path` property is represented as a `post_logout_redirect_uri` query parameter which will not be recognized by the providers which do not support this specification.

Reviewer: ```suggestion
According to the https://openid.net/specs/openid-connect-session-1_0.html#RPLogout[RP-initiated logout] specification, the `quarkus.oidc.logout.post-logout-path` property is represented as a `post_logout_redirect_uri` query parameter, which is not recognized by the providers that do not support this specification.
```

# For example, if 'client_id' is provided then a valid logout URI should be set as Auth0 Application property, without it - as Auth0 Tenant property.
quarkus.oidc.logout.extra-params.client_id=${quarkus.oidc.client-id}
----
--

[[back-channel-logout]]
===== Back-channel logout

The OIDC provider can force the logout of all applications using the authentification data: this is called back-channel logout.
In this case the OIDC will call a specific URL from each application to trigger that logout.

link:https://openid.net/specs/openid-connect-backchannel-1_0.html[Back-Channel Logout] is used by OpenID Connect providers to log out the current user from all the applications this user is currently logged in, bypassing the user agent.

You can configure Quarkus to support `Back-Channel Logout` as follows:

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
quarkus.oidc.client-id=frontend
quarkus.oidc.credentials.secret=secret
quarkus.oidc.application-type=web-app

quarkus.oidc.logout.backchannel.path=/back-channel-logout
----

> Absolute `Back-Channel Logout` URL is calculated by adding `quarkus.oidc.back-channel-logout.path` to the current endpoint URL, for example, `http://localhost:8080/back-channel-logout`. You will need to configure this URL in the Admin Console of your OpenID Connect Provider.

Reviewer: ```suggestion
The absolute `back-channel logout` URL is calculated by adding `quarkus.oidc.back-channel-logout.path` to the current endpoint URL, for example, `http://localhost:8080/back-channel-logout`. 
You will need to configure this URL in the admin console of your OIDC provider.
```

> Note that you will also need to configure a token age property for the logout token verification to succeed if your OpenID Connect Provider does not set an expiry claim in the current logout token, for example, `quarkus.oidc.token.age=10S` sets a number of seconds that must not elapse since the logout token's `iat` (issued at) time to 10.

Reviewer: ```suggestion
You will also need to configure a token age property for the logout token verification to succeed if your OpenID Connect Provider does not set an expiry claim in the current logout token.
For example, set `quarkus.oidc.token.age=10S` to ensure that no more than 10 seconds elapse since the logout token's `iat` (issued at) time.
```

[[front-channel-logout]]
===== Front-channel logout

link:https://openid.net/specs/openid-connect-frontchannel-1_0.html[Front-Channel Logout] can be used to logout the current user directly from the user agent (e.g. its browser).
It is like <<back-channel-logout,Back-channel logout>> but the logout steps are executed by the user agent like the browser and not in the background by the OIDP provider.
This option is rarely used.

You can configure Quarkus to support `Front-Channel Logout` as follows:

[source,properties]
----
quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
quarkus.oidc.client-id=frontend
quarkus.oidc.credentials.secret=secret
quarkus.oidc.application-type=web-app

quarkus.oidc.logout.frontchannel.path=/front-channel-logout
----

This path will be compared against the current request's path and the user will be logged out if these paths match.

[[local-logout]]
===== Local logout

> <<user-initiated-logout,User-initiated logout>> will log the person out from the OIDC provider, if it is used as single sign on, it might not be what you desire.

Reviewer: ```suggestion
<<user-initiated-logout,User-initiated logout>> will log the user out of the OIDC provider, if it is used as single-sign on, it might not be what you require.
```

> If for example your OIDC provider is Google, you will be logged out from Google and its services.

Reviewer: ```suggestion
If, for example, your OIDC provider is Google, you will be logged out from Google and its services.
```

Instead the user might just want to log out of that specific application.

> Another use case if the OIDC provider does not have a logout endpoint.

Reviewer: ```suggestion
Another use case might be when the OIDC provider does not have a logout endpoint.
```

> You can support a local logout with the help of the <<oidc-session,OidcSession>> which only clears the local session cookie, for example:

Reviewer: ```suggestion
By using <<oidc-session,OidcSession>>, you can support a local logout, which means that only the local session cookie is cleared, as shown in the following example:
```

[source,java]
----
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;

import io.quarkus.oidc.OidcSession;

@Path("/service")
public class ServiceResource {

    @Inject
    OidcSession oidcSession;

    @GET
    @Path("logout")
    public String logout() {
        oidcSession.logout().await().indefinitely();
        return "You are logged out".
    }
----

[[oidc-session]]

> ====== Using OidcSession for local logout

Reviewer: ```suggestion
====== Using `OidcSession` for local logout
```

> `io.quarkus.oidc.OidcSession` is a wrapper around the current `IdToken`. It can help to perform a <<local-logout, Local logout>>, retrieve the current session's tenant identifier and check when the session will expire. More useful methods will be added to it over time.

Reviewer: ```suggestion
`io.quarkus.oidc.OidcSession` is a wrapper around the current `IdToken`, which can help to perform a <<local-logout, Local logout>>, retrieve the current session's tenant identifier, and check when the session will expire. 
More useful methods will be added to it over time.
```

[[session-management]]

> ==== Expiration via Session management

Reviewer: ```suggestion
==== Expiration through session management
```

Reviewer 2: @michelle-purcell I did not notice Emmanuel changed this section title, it should remain as `Session management`, adding `Expiration` in front introduces a new meaning

Reviewer: Fixed :+1:

> By default, the logout is based on the expiration time of the ID Token issued by the OpenID Connect Provider. When the ID Token expires, the current user session at the Quarkus endpoint is invalidated and the user is redirected to the OpenID Connect Provider again to authenticate. If the session at the OpenID Connect Provider is still active, users are automatically re-authenticated without having to provide their credentials again.

Reviewer: ```suggestion
By default, logout is based on the expiration time of the ID token issued by the OIDC provider. When the ID token expires, the current user session at the Quarkus endpoint is invalidated and the user is redirected to the OIDC provider again to authenticate. 
If the session at the OIDC provider is still active, users are automatically re-authenticated without needing to provide their credentials again.
```

> The current user session may be automatically extended by enabling a `quarkus.oidc.token.refresh-expired` property. If it is set to `true` then when the current ID Token expires a Refresh Token Grant will be used to refresh ID Token as well as Access and Refresh Tokens.

Reviewer: ```suggestion
The current user session can be automatically extended by enabling the `quarkus.oidc.token.refresh-expired` property. 
If set to `true`, when the current ID token expires, a refresh token grant will be used to refresh the ID token as well as access and refresh tokens.
```

[TIP]
--

> If you have a xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[Single Page Application for Service Applications] where your OpenID Connect Provider script such as `keycloak.js` is managing an authorization code flow then that script will also control the SPA authentication session lifespan.

Reviewer: ```suggestion
If you have a xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[single page application for service applications] where your OIDC provider script such as `keycloak.js` is managing an authorization code flow then that script will also control the SPA authentication session lifespan.
```

--

> If you work with a Quarkus OIDC `web-app` application then it is Quarkus OIDC Code Authentication mechanism which is managing the user session lifespan.

Reviewer: ```suggestion
If you work with a Quarkus OIDC `web-app` application then it is the Quarkus OIDC code authentication mechanism that is managing the user session lifespan.
```

> To use the refresh token, you should carefully configure the various lifespan of the ID token, the refresh token, the cookie, etc so they work in tandem.

Reviewer: ```suggestion
To use the refresh token, you should carefully configure the various lifespan of the ID token, the refresh token, the cookie, and any other elements needed, so that they work in tandem.
```

Reviewer 2: I'd like to review this sentence, on its own it does not make a lot of sense (meaning I likely created it in the first place :-) )

Reviewer: @sberyozkin - No problem.  Let me know if you need my input on the rewrite. Thanks.

Reviewer 2: @michelle-purcell Thanks, for now, I simplified it to `To use the refresh token, you should carefully configure session cookie age.` with the instructions following

If the session at the OIDC provider is still active, users are automatically re-authenticated without needing to provide their credentials again.

The current user session can be automatically extended by enabling the `quarkus.oidc.token.refresh-expired` property. 
If set to `true`, when the current ID token expires, a refresh token grant will be used to refresh the ID token as well as access and refresh tokens.

[TIP]
--
If you have a xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[single page application for service applications] where your OIDC provider script such as `keycloak.js` is managing an authorization code flow then that script will also control the SPA authentication session lifespan.
--

If you work with a Quarkus OIDC `web-app` application, then it is the Quarkus OIDC code authentication mechanism that is managing the user session lifespan.

To use the refresh token, you should carefully configure the various lifespan of the ID token, the refresh token, the cookie, etc so they work in tandem.

> The so called session age should be longer than the ID token lifespan and close to or equal to the refresh token lifespan.

Reviewer 3: ```suggestion
The session age should be longer than the ID token lifespan and close to or equal to the refresh token lifespan.
```

> The session age is calculated by adding the lifespan value of the current IDToken and the values of the `quarkus.oidc.authentication.session-age-extension` and `quarkus.oidc.token.lifespan-grace` properties. Of the last two properties only `quarkus.oidc.authentication.session-age-extension` should be used to significantly extend the session lifespan if required since `quarkus.oidc.token.lifespan-grace` is only meant for taking some small clock skews into consideration.

Reviewer 3: ```suggestion
You calculate the session age by adding the lifespan value of the current ID token and the values of the `quarkus.oidc.authentication.session-age-extension` and `quarkus.oidc.token.lifespan-grace` properties. 
[NOTE]
--
You use only the `quarkus.oidc.authentication.session-age-extension` property to significantly extend the session lifespan, if required. 
You use the `quarkus.oidc.token.lifespan-grace` property only for taking some small clock skews into consideration.
--

Reviewer 2: @sheilamjones making it a `TIP` makes sense. The clock skew is a difference in clock setups on various computers, it is a well known technical term for it, so I suppose we can leave it as is

> When the currently authenticated user returns to the protected Quarkus endpoint and the ID token associated with the session cookie has expired then, by default, the user will be auto-redirected to the OIDC Authorization endpoint to re-authenticate. Most likely the OIDC provider will challenge the user again though not necessarily if the session between the user and this OIDC provider is still active which may happen if it is configured to last longer than the ID token.

Reviewer 3: ```suggestion
When the current authenticated user returns to the protected Quarkus endpoint and the ID token associated with the session cookie has expired, then, by default, the user is automatically redirected to the OIDC Authorization endpoint to re-authenticate. 
The OIDC provider might challenge the user again if the session between the user and this OIDC provider is still active, which might happen if the session is configured to last longer than the ID token.
```

> If the `quarkus.oidc.token.refresh-expired` is set to `true`, then the expired ID token (as well as the access token) will be refreshed using the refresh token returned with the initial authorization code grant response. This refresh token may also be recycled (refreshed) itself as part of this process. As a result the new session cookie will be created and the session will be extended.

Reviewer 3: ```suggestion
If the `quarkus.oidc.token.refresh-expired` is set to `true`, then the expired ID token (as well as the access token) is refreshed by using the refresh token returned with the initial authorization code grant response. 
This refresh token might also be recycled (refreshed) itself as part of this process. 
As a result, the new session cookie is created and the session is extended.
```

[NOTE]
--

> `quarkus.oidc.authentication.session-age-extension` can be important when dealing with expired ID tokens, when the user is not very active. If the ID token expires, then the session cookie may not be returned to the Quarkus endpoint during the next user request as the cookie lifespan would ahve elapsed. Quarkus will assume it is the first authentication request. Set `quarkus.oidc.authentication.session-age-extension` to be _reasonably_ long for your barely active users and in accordance with your security policies.

Reviewer 3: ```suggestion
In instances where the user is not very active, you can use the `quarkus.oidc.authentication.session-age-extension` property to help handle expired ID tokens. 
If the ID token expires, the session cookie might not be returned to the Quarkus endpoint during the next user request as the cookie lifespan would have elapsed. 
Quarkus assumes that this request is the first authentication request. 
Set `quarkus.oidc.authentication.session-age-extension` to be _reasonably_ long for your barely-active users and in accordance with your security policies.
```

--

You can go one step further and proactively refresh ID tokens or access tokens that are about to expire.
Set `quarkus.oidc.token.refresh-token-time-skew` to the value you want to anticipate the refresh. If, during the current user request, it is calculated that the current ID token will expire within this `quarkus.oidc.token.refresh-token-time-skew` then it will be refreshed and the new session cookie will be created. This property should be set to a value which is less than the ID token lifespan; the closer it is to this lifespan value the more often the ID token will be refreshed.

> You can have this process further optimized by having a simple JavaScript function periodically emulating the user activity by pinging your Quarkus endpoint thus minimizing the window during which the user may have to be re-authenticated.

Reviewer 3: ```suggestion
You can further optimize this process by having a simple JavaScript function ping your Quarkus endpoint to periodically emulate the user activity, which minimizes the time frame during which the user might have to be re-authenticated.
```

> Note this user session can not be extended forever - the returning user with the expired ID token will have to re-authenticate at the OIDC provider endpoint once the refresh token has expired.

Reviewer 3: ```suggestion
[NOTE]
--
You cannot extend this user session indefinitely.
The returning user with the expired ID token will have to re-authenticate at the OIDC provider endpoint once the refresh token has expired.
--
```

[[oauth2]]
=== Integration with GitHub and non-OIDC OAuth2 providers

Some well known providers such as GitHub or LinkedIn are not OpenID Connect but OAuth2 providers which support the `authorization code flow`, for example, link:https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps[GitHub OAuth2] and link:https://docs.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow[LinkedIn OAuth2].
Remember, OIDC is built on top of OAuth2.

> The main difference between OpenID Connect and OAuth2 providers is that OpenID Connect providers return an `ID Token` representing a user authentication, in addition to the standard authorization code flow `access` and `refresh` tokens returned by `OAuth2` providers.

Reviewer 3: ```suggestion
The main difference between OpenID Connect and OAuth2 providers is that OpenID Connect providers return an `ID Token` that represents a user authentication, in addition to the standard authorization code flow `access` and `refresh` tokens returned by `OAuth2` providers.
```

OAuth2 providers such as GitHub do not return `IdToken`, the fact of the user authentication is implicit and is indirectly represented by the `access` token which represents an authenticated user authorizing the current Quarkus `web-app` application to access some data on behalf of the authenticated user.

> While in the OIDC case, we validate the ID Token as proof of the authentication validity, in the raw OAuth2 case, we need to validate the access token.

Reviewer 3: ```suggestion
For OIDC, you validate the ID token as proof of authentication validity whereas in the case of raw OAuth2, you validate the access token.
```

Reviewer 3: Question: Do we need to say "raw OAuth2" here? Could we perhaps change to "OAuth2 only" perhaps?

Reviewer 2: Hi @sheilamjones Yeah, makes sense to drop a `raw` qualifier, thanks. Updated your comment and will apply

> This is done by subsequently calling an endpoint that requires the access token and that typically returns user information.

Reviewer 3: Question: What is meant by "subsequently" here? Also, is "typically" needed here? Are there instances where the user information is not returned? Could this sentence be adjusted as follows perhaps"
"You validate the access token by calling an endpoint that requires the token. 
This action returns the user information."

Reviewer 2: Hi @sheilamjones It is probably the right balance there, as it tries to capture various variations without going into too many details. `subsequently` in this context means that after the code flow has completed, a follow up request is done to fetch more information, `typically` it would be a user information but in `OAuth2`, as opposed to `OIDC`, `UserInfo` is undefined, so whatever endpoint is available, like a user profile or something close enough that cab be used to indirectly validate the access token, will do...
So we should probably leave it as is

> This very much the same as the OIDC <<user-info,UserInfo>> approach which is fetched by the OIDC flow on your behalf.

Reviewer 3: Suggestion: "This approach is similar to the OIDC <<user-info,UserInfo>> approach, which is fetched by the OIDC flow on your behalf."

Reviewer 2: @sheilamjones I typed it like this: `This approach is similar to the OIDC <<user-info,UserInfo>> approach, with `UserInfo` fetched by Quarkus OIDC on your behalf.` which reads a bit more precise as opposed to `by OIDC flow`

> For example, when working with GitHub, the Quarkus endpoint can acquire an `access` token which will allow it to request a GitHub profile of the current user.

Reviewer 3: Suggestion: Some ambiguity in the original sentence due to the use of "it". Is the above suggestion correct?

> For example, when working with GitHub, the Quarkus endpoint can acquire an `access` token which will allow it to request a GitHub profile of the current user.

Reviewer 2: @sheilamjones Yep, thanks

> To support the integration with such OAuth2 servers, `quarkus-oidc` needs to be configured a bit differently to allow the authorization code flow responses without `IdToken`: `quarkus.oidc.authentication.id-token-required=false`.

Reviewer 3: Question: Is it meant here that the authorization code flow responses are specified without the `IdToken` being specified with the property `quarkus.oidc.authentication.id-token-required=false`?

[NOTE]
--

> Even though you will configure the extension to support the authorization code flows without `IdToken`, an internal `IdToken` will be generated to standardize the way `quarkus-oidc` operates: an `IdToken` is used to support the authentication session and to avoid redirecting the user to the provider such as GitHub on every request. In this case the session lifespan is set to 5 minutes which can be extended further as described in the <<session-management,session management>> section.

Reviewer 3: ```suggestion
Even though you configure the extension to support the authorization code flows without `IdToken`, an internal `IdToken` is generated to standardize the way `quarkus-oidc` operates. 
You use an `IdToken` to support the authentication session and to avoid redirecting the user to the provider,  such as GitHub, on every request. 
In this case, the session lifespan is set to 5 minutes, which you can can extend further as described in the <<session-management,session management>> section.
```

> This simplifies how you handle an application supporting multiple OIDC providers.

Reviewer 3: ```suggestion
This simplifies how you handle an application that supports multiple OIDC providers.
```

--

The next step is to ensure that the returned access token can be useful and is valid to the current Quarkus endpoint.

> The first way is to call the OAuth2 provider introspection endpoint by configuring `quarkus.oidc.introspection-path` if the provider offers such an endpoint.

Reviewer 3: ```suggestion
The first way is to call the OAuth2 provider introspection endpoint by configuring `quarkus.oidc.introspection-path`, if the provider offers such an endpoint.
```

> In this case you can use the access token as a source of role using `quarkus.oidc.roles.source=accesstoken`.

Reviewer 3: Question: Not quite sure what is meant here by "as a source of role". Would it be correct to say something like" "In this case, you can specify `access token` as the value for the `quarkus.oidc.roles.source` property, that is, `quarkus.oidc.roles.source=access token`.

Reviewer 2: @sheilamjones It should be a `a source of roles` instead of `a source of role`, thanks for highlighting it, I'd limit it though to this minor change only. For the code flow, the ID token typically contains roles, but there could be some variations (access token can be a source of this information, or userinfo)

Reviewer 2: @sheilamjones I changed `role` to `roles` locally

> If no introspection endpoint is present, your next best step is to request <<user-info,UserInfo>> from the provider as it will at least validate the access token. Use `quarkus.oidc.token.verify-access-token-with-user-info=true`.

Reviewer 3: ```suggestion
If no introspection endpoint is present, you can attempt instead to request <<user-info,UserInfo>> from the provider as it will at least validate the access token. 
To do so, specify `quarkus.oidc.token.verify-access-token-with-user-info=true`.
```

> You also need to set the `quarkus.oidc.user-info-path` property to a URL endpoint fetching the user info (or at least to an endpoint protected by the access token).

Reviewer 3: ```suggestion
You also need to set the `quarkus.oidc.user-info-path` property to a URL endpoint that fetches the user info (or at least to an endpoint protected by the access token).
```

> For GitHub, since it does not have an introspection endpoint, we use the last method.

Reviewer 3: Question: To help avoid ambiguity here and to avoid possible confusion after future updates to the doc, rather than "the last method", could you be more specific here and name the method?

Reviewer 2: @sheilamjones Sure, good catch, can you please propose something like `For GitHub, since it does not have an introspection endpoint, requesting the UserInfo is required.` where what is needed to `request UserInfo` is described just above, so users will know

Reviewer 2: @sheilamjones Done as suggested above

[NOTE]
--

> Requiring <<user-info,UserInfo>> involves making a remote call on every request - therefore you may want to consider caching `UserInfo` data, see xref:security-oidc-bearer-token-authentication-concept.adoc#token-introspection-userinfo-cache[Token Introspection and UserInfo cache] for more details.

Reviewer 3: ```suggestion
Requiring <<user-info,UserInfo>> involves making a remote call on every request.
Therefore, you might want to consider caching `UserInfo` data.
For more information, see the xref:security-oidc-bearer-token-authentication-concept.adoc#token-introspection-userinfo-cache[Token Introspection and UserInfo cache] section of the "OpenID Connect (OIDC) Bearer authentication" guide.
```

Alternatively, you may want to request that `UserInfo` is embedded into the internal generated `IdToken` with the `quarkus.oidc.cache-user-info-in-idtoken=true` property - the advantage of this approach is that by default no cached `UserInfo` state will be kept with the endpoint - instead it will be stored in a session cookie. You may also want to consider encrypting `IdToken` in this case if `UserInfo` contains sensitive data. For more information, see <<token-state-manager,Encrypt tokens with TokenStateManager>>.
--

> OAuth2 servers may not support a well-known configuration endpoint in which case the discovery has to be disabled and the authorization, token, and introspection and/or userinfo endpoint paths have to be configured manually.

Reviewer 3: ```suggestion
OAuth2 servers might not support a well-known configuration endpoint.
In this case, you must disable the discovery and configure the authorization, token, and introspection and the userinfo endpoint paths manually.
```

Reviewer 3: We avoid using an "and/or" conjunction in docs due to possible ambiguity or issues with translation. Instead, depending on the context, we use one of the following constructions: a and b, a or b, or a, b, or both. Is my above rewording correct in this instance?

> This can look a bit daunting but Quarkus simplifies your life for well known OIDC or OAuth2 providers like Apple, Facebook, GitHub, Google, Microsoft, Spotify and Twitter. Use the `quarkus.oidc.provider` property.

Reviewer 3: ```suggestion
For well-known OIDC or OAuth2 providers, like Apple, Facebook, GitHub, Google, Microsoft, Spotify, and Twitter, Quarkus can help simplify your processes.
Use the `quarkus.oidc.provider` property.
```

Reviewer 3: Update suggestion was to make the statement a bit more direct.

        return Uni.createFrom().item(identity);
    }

}
----

> Now, the following code will work when the user is signing in into your application with both Google or GitHub:

Reviewer 3: ```suggestion
Now, the following code will work when the user signs into your application by using Google or GitHub:
```

[source,java]
----
import jakarta.inject.Inject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;

import io.quarkus.security.Authenticated;
import io.quarkus.security.identity.SecurityIdentity;

@Path("/service")
@Authenticated
public class TokenResource {

    @Inject
    SecurityIdentity identity;

    @GET
    @Path("/google")
    @Produces("application/json")
    public String getUserName() {
        return identity.getPrincipal().getName();
    }

    @GET
    @Path("/github")
    @Produces("application/json")
    public String getUserName() {
        return identity.getPrincipal().getUserName();
    }
}
----

> Possibly a simpler alternative is to inject both `@IdToken JsonWebToken` and `UserInfo` and use `JsonWebToken` when dealing with the providers returning `IdToken` and `UserInfo` - with the providers which do not return `IdToken`.

Reviewer 3: Question: Maybe it's just me, but I'm not sure what is meant here due to the part after the hyphen at the end of the sentence ".....- with the providers which do not return `IdToken`? Can this be rephrased perhaps to remove  the hyphen?

Reviewer 2: Done

> The last important point is to make sure the callback path you enter in the GitHub OAuth application configuration matches the endpoint path where you'd like the user be redirected to after a successful GitHub authentication and application authorization, in this case it has to be set to `http:localhost:8080/github/userinfo`.

Reviewer 3: ```suggestion
The last important point is to make sure the callback path you enter in the GitHub OAuth application configuration matches the endpoint path where you want to redirect the user after a successful GitHub authentication and application authorization.
In this case, you must set the callback path to `http:localhost:8080/github/userinfo`.
```

Reviewer 3: Suggestion: I would also suggest removing "important point" here because it might imply that the other point are not/less important. What about rephrasing to "You must ensure that the callback path you enter in the GitHub....." and if it is of particular importance to highlght, this section could be encoded in an [IMPORTANT] block?

Reviewer 2: @sheilamjones Sure, can you please propose a change I can commit ?

Reviewer 2: @sheilamjones Updated as you proposed

=== Listening to important authentication events

> One can register `@ApplicationScoped` bean which will observe important OIDC authentication events. The listener will be updated when a user has logged in for the first time or re-authenticated, as well as when the session has been refreshed. More events may be reported in the future. For example:

Reviewer 3: ```suggestion
You can register the `@ApplicationScoped` bean which will observe important OIDC authentication events.
When a user logs in for the first time or reauthenticates, or refreshes the session the listener is updated.  
In the future, more events might be reported. 
For example:
```

[source, java]
----
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.event.Observes;

import io.quarkus.oidc.IdTokenCredential;
import io.quarkus.oidc.SecurityEvent;
import io.quarkus.security.identity.AuthenticationRequestContext;
import io.vertx.ext.web.RoutingContext;

@ApplicationScoped
public class SecurityEventListener {

    public void event(@Observes SecurityEvent event) {
        String tenantId = event.getSecurityIdentity().getAttribute("tenant-id");
        RoutingContext vertxContext = event.getSecurityIdentity().getAttribute(RoutingContext.class.getName());
        vertxContext.put("listener-message", String.format("event:%s,tenantId:%s", event.getEventType().name(), tenantId));
    }
}
----

=== Propagating tokens to downstream services

> For information about Authorization Code Flow access token propagation to the downstream services, see the xref:security-openid-connect-client-reference.adoc#token-propagation[Token Propagation] section.

Reviewer 3: ```suggestion
For information about Authorization Code Flow access token propagation to downstream services, see the xref:security-openid-connect-client-reference.adoc#token-propagation[Token Propagation] section.
```

== Integration considerations

> Your OIDC backed application integrates in an environment such as calling from a single page application, using well known OIDC providers etc. 

Reviewer 2: @sheilamjones Yeah, I agree. Let me think

This section discusses these considerations.

=== Single-page applications

> Check if implementing SPAs the way it is suggested in the xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[Single-page Applications for Service Applications] section can meet your requirements.

Reviewer 3: ```suggestion
Check if implementing single-page applications (SPAs) the way it is suggested in the xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[Single-page Applications for Service Applications] section can meet your requirements.
```

> If you prefer to use SPA and JavaScript API such as `Fetch` or `XMLHttpRequest`(XHR) with Quarkus web applications, be aware that OpenID Connect Providers may not support CORS for Authorization endpoints where the users are authenticated after a redirect from Quarkus. This will lead to authentication failures if the Quarkus application and the OpenID Connect Provider are hosted on the different HTTP domains/ports.

Reviewer 3: ```suggestion
If you prefer to use SPAs and JavaScript APIs such as `Fetch` or `XMLHttpRequest`(XHR) with Quarkus web applications, be aware that OpenID Connect providers might not support cross-origin resource sharing (CORS) for authorization endpoints where the users are authenticated after a redirect from Quarkus. 
This will lead to authentication failures if the Quarkus application and the OpenID Connect provider are hosted on different HTTP domains, ports, or both.
```

> In such cases, set the `quarkus.oidc.authentication.java-script-auto-redirect` property to `false` which will instruct Quarkus to return a `499` status code and `WWW-Authenticate` header with the `OIDC` value. The browser script also needs to be updated to set `X-Requested-With` header with the `JavaScript` value and reload the last requested page in case of `499`, for example:

Reviewer 3: ```suggestion
In such cases, set the `quarkus.oidc.authentication.java-script-auto-redirect` property to `false`, which will instruct Quarkus to return a `499` status code and a `WWW-Authenticate` header with the `OIDC` value. 
You must also update the browser script to set the `X-Requested-With` header with the `JavaScript` value and reload the last requested page in case of a `499` status code.
For example:
```

[source,javascript]
----
Future<void> callQuarkusService() async {
    Map<String, String> headers = Map.fromEntries([MapEntry("X-Requested-With", "JavaScript")]);

    await http
        .get("https://localhost:443/serviceCall")
        .then((response) {
            if (response.statusCode == 499) {
                window.location.assign("https://localhost.com:443/serviceCall");
            }
         });
  }
----

=== Cross-origin resource sharing

If you plan to consume this application from a Single-page application running on a different domain, you will need to configure CORS (Cross-Origin Resource Sharing). For more information, see the xref:http-reference.adoc#cors-filter[HTTP CORS documentation].

=== Calling Cloud provider services

==== Google Cloud

> You can have Quarkus OIDC `web-app` applications access **Google Cloud services** such as **BigQuery** on behalf of the currently authenticated users who have enabled OpenID Connect (Authorization Code Flow) permissions to such services in their Google Developer Consoles.

Reviewer 3: ```suggestion
You can have Quarkus OIDC `web-app` applications access **Google Cloud services** such as **BigQuery** on behalf of the currently-authenticated users who have enabled OpenID Connect (Authorization Code Flow) permissions to such services in their Google developer consoles.
```

import jakarta.enterprise.event.Observes;

import io.quarkus.oidc.IdTokenCredential;
import io.quarkus.oidc.SecurityEvent;
import io.quarkus.security.identity.AuthenticationRequestContext;
import io.vertx.ext.web.RoutingContext;

@ApplicationScoped
public class SecurityEventListener {

    public void event(@Observes SecurityEvent event) {
        String tenantId = event.getSecurityIdentity().getAttribute("tenant-id");
        RoutingContext vertxContext = event.getSecurityIdentity().getAttribute(RoutingContext.class.getName());
        vertxContext.put("listener-message", String.format("event:%s,tenantId:%s", event.getEventType().name(), tenantId));
    }
}
----

=== Propagating tokens to downstream services

For information about Authorization Code Flow access token propagation to downstream services, see the xref:security-openid-connect-client-reference.adoc#token-propagation[Token Propagation] section.

== Integration considerations

Your OIDC backed application integrates in an environment such as calling from a single page application, using well known OIDC providers etc. 
This section discusses these considerations.

=== Single-page applications

> Check if implementing single-page applications (SPAs) the way it is suggested in the xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[Single-page Applications for Service Applications] section can meet your requirements.

Reviewer 3: ```suggestion
Check if implementing single-page applications (SPAs) the way it is suggested in the xref:security-oidc-bearer-token-authentication-concept.adoc#single-page-applications[Single-page applications] section of the "OpenID Connect (OIDC) Bearer authentication" guide meets your requirements.
```

quarkus.log.category."io.quarkus.oidc.runtime.OidcRecorder".min-level=TRACE
----

> You can also from `quarkus dev` console hit `j` to change the application global log level.

Reviewer 3: ```suggestion
To change the application global log level, you can also press `j` from `quarkus dev` console.
```
